pkg/networkservice/chains/nsmgr/server.go

```diff
@@ -14,65 +14,110 @@
 
-// Package nsmgr provides a Network Service Manager (nsmgr), but interface and implementation
+// Package nsmgr provides a Network Service Manager (nsmgrServer), but interface and implementation
 package nsmgr
 
 import (
+	"context"
+	"net/url"
+
+	"github.com/networkservicemesh/sdk/pkg/networkservice/chains/nsmgr/authz"
+	"github.com/networkservicemesh/sdk/pkg/networkservice/common/connect"
+	"github.com/networkservicemesh/sdk/pkg/registry/nsmreg"
+
 	"github.com/networkservicemesh/api/pkg/api/networkservice"
 	"github.com/networkservicemesh/api/pkg/api/registry"
-	"google.golang.org/grpc"
 
 	"github.com/networkservicemesh/sdk/pkg/networkservice/chains/client"
-	"github.com/networkservicemesh/sdk/pkg/networkservice/chains/endpoint"
-	"github.com/networkservicemesh/sdk/pkg/networkservice/common/connect"
+	"github.com/networkservicemesh/sdk/pkg/networkservice/common/authorize"
 	"github.com/networkservicemesh/sdk/pkg/networkservice/common/discover"
 	"github.com/networkservicemesh/sdk/pkg/networkservice/common/localbypass"
 	"github.com/networkservicemesh/sdk/pkg/networkservice/common/roundrobin"
 	"github.com/networkservicemesh/sdk/pkg/networkservice/core/adapters"
 	adapter_registry "github.com/networkservicemesh/sdk/pkg/registry/core/adapters"
-	chain_registry "github.com/networkservicemesh/sdk/pkg/registry/core/chain"
+	"github.com/networkservicemesh/sdk/pkg/registry/memory"
 	"github.com/networkservicemesh/sdk/pkg/tools/addressof"
+
+	"google.golang.org/grpc"
+
+	"github.com/networkservicemesh/sdk/pkg/networkservice/chains/endpoint"
+	"github.com/networkservicemesh/sdk/pkg/tools/callback"
 	"github.com/networkservicemesh/sdk/pkg/tools/token"
 )
 
 // Nsmgr - A simple combintation of the Endpoint, registry.NetworkServiceRegistryServer, and registry.NetworkServiceDiscoveryServer interfaces
 type Nsmgr interface {
 	endpoint.Endpoint
+
 	registry.NetworkServiceRegistryServer
 	registry.NetworkServiceDiscoveryServer
+	registry.NsmRegistryServer
+
+	callback.CallbackServiceServer
 }
 
-type nsmgr struct {
+type nsmgrServer struct {
 	endpoint.Endpoint
+
 	registry.NetworkServiceRegistryServer
 	registry.NetworkServiceDiscoveryServer
+	registry.NsmRegistryServer
+
+	storage memory.Storage
+	callback.CallbackServiceServer
 }
 
 // NewServer - Creates a new Nsmgr
 //           name - name of the Nsmgr
 //           authzServer - authorization server chain element
-//           registryCC - client connection to reach the upstream registry
-func NewServer(name string, authzServer networkservice.NetworkServiceServer, tokenGenerator token.GeneratorFunc, registryCC grpc.ClientConnInterface, dialOptions ...grpc.DialOption) Nsmgr {
-	rv := &nsmgr{}
+//           registryCC - client connection to reach the upstream registry, could be nil, in this case only in memory storage will be used.
+func NewServer(manager *registry.NetworkServiceManager, tokenGenerator token.GeneratorFunc, registryCC grpc.ClientConnInterface) Nsmgr {
```

Comment:
Could we rename manager here to something like nsmRegistry or nsmRegistryPtr to make it clearer?  My first scan of this reading the variable name lead me to an incorrect conclusion.

Reply:
Changed to nsmRegistration, since it Network Service Manager registration, will need to update nsm registry chain elements to use same name I suppose.